[5.8] Provide retry callback with the attempt count

AJenbo opened the pull request:
This provides the call back with a attempt count, this can be helpful for function that need to modify there behavior with each attempt, or log it some how. Existing code should be unaffected since there was not parameter earlier.

AJenbo committed:
Provide retry callback with the attempt count

This provides the call back with a attempt count, this can be helpful
for function that need to modify there behavior with each attempt, or
log it some how.

BrandonSurowiec commented:
This method was written using `goto` for a very specific reason. 

Check out this PR that contains links to threads about it:
https://github.com/laravel/framework/pull/15973

AJenbo commented:
Thanks for the quick review. I have rewritten the code so that it keeps the `goto` and added a test for retry that test this and existing features.

I would add that the `goto` feel like an premature optimization, with the addition of `$attempt++` this would also mean that it saves a jump or two at the cost of an additional `$times--`, also the php code is 7 lines longer.
